state/secrets.go

```diff
@@ -1836,3 +1864,182 @@ func (w *secretsRotationWatcher) loop() (err error) {
 		}
 	}
 }
+
+// WatchSecretRevisionsExpiryChanges returns a watcher for expiry time updates to
+// secret revisions with the specified owner.
+func (st *State) WatchSecretRevisionsExpiryChanges(owner string) SecretsTriggerWatcher {
+	return newSecretsExpiryWatcher(st, owner)
+}
+
+type expiryWatcherDetails struct {
+	txnRevNo   int64
+	uri        *secrets.URI
+	revision   int
+	willExpire bool
+}
+
+type secretsExpiryWatcher struct {
+	commonWatcher
+	out chan []corewatcher.SecretTriggerChange
+
+	owner string
+	known map[string]expiryWatcherDetails
+}
+
+func newSecretsExpiryWatcher(backend modelBackend, owner string) *secretsExpiryWatcher {
+	w := &secretsExpiryWatcher{
+		commonWatcher: newCommonWatcher(backend),
+		out:           make(chan []corewatcher.SecretTriggerChange),
+		known:         make(map[string]expiryWatcherDetails),
+		owner:         owner,
+	}
+	w.tomb.Go(func() error {
+		defer close(w.out)
+		return w.loop()
+	})
+	return w
+}
+
+// Changes returns a channel that will receive changes when the next rotate time
+// for a secret changes.
+func (w *secretsExpiryWatcher) Changes() corewatcher.SecretTriggerChannel {
+	return w.out
+}
+
+func (w *secretsExpiryWatcher) initial() ([]corewatcher.SecretTriggerChange, error) {
+	var details []corewatcher.SecretTriggerChange
+
+	var doc secretRevisionDoc
+	secretRevisionCollection, closer := w.db.GetCollection(secretRevisionsC)
+	defer closer()
+
+	iter := secretRevisionCollection.Find(bson.D{{"owner-tag", w.owner}}).Iter()
+	for iter.Next(&doc) {
+		uriStr, _ := splitSecretChange(w.backend.localID(doc.DocID))
+		uri, err := secrets.ParseURI(uriStr)
+		if err != nil {
+			_ = iter.Close()
+			return nil, errors.Annotatef(err, "invalid secret URI %q", uriStr)
+		}
+		willExpire := doc.ExpireTime != nil
+		w.known[doc.DocID] = expiryWatcherDetails{
+			txnRevNo:   doc.TxnRevno,
+			uri:        uri,
+			revision:   doc.Revision,
+			willExpire: willExpire,
+		}
+		if !willExpire {
+			continue
+		}
+		details = append(details, corewatcher.SecretTriggerChange{
+			URI:             uri,
+			Revision:        doc.Revision,
+			NextTriggerTime: doc.ExpireTime.UTC(),
+		})
+	}
+	return details, errors.Trace(iter.Close())
+}
+
+func (w *secretsExpiryWatcher) merge(details []corewatcher.SecretTriggerChange, change watcher.Change) ([]corewatcher.SecretTriggerChange, error) {
+	changeID := change.Id.(string)
+	knownDetails, known := w.known[changeID]
+
+	doc := secretRevisionDoc{}
+	if change.Revno >= 0 {
+		secretRevisionCollection, closer := w.db.GetCollection(secretRevisionsC)
+		defer closer()
+		err := secretRevisionCollection.Find(bson.D{{"_id", change.Id}, {"owner-tag", w.owner}}).One(&doc)
+		if err != nil && err != mgo.ErrNotFound {
+			return nil, errors.Trace(err)
+		}
+		// Changed but no longer in the collection so ignore.
+		if err != nil {
+			return details, nil
+		}
+	} else if known {
+		// Record deleted.
```

Comment:
revno is never 0 - it will be -1 here as checked above for a non deleted record
because revno -1 means the record has already been deleted, there's no point doing a db query